Update Course: Form refactoring, including datetime-picker

thilo-behnke opened the pull request:
This PR was migrated from https://github.com/ls1intum/ArTEMiS/pull/166.

### Checklist
- [x] I've run `yarn run webpack:build:main` from the root directory to see that the project builds without errors.
- [x] I've removed unnecessary whitespace changes.
- [ ] I've updated the documentation and models if necessary.
- [ ] I've tested the changes and all related features on the Artemis test server.
- [ ] I've added (end-to-end) test cases for all new functionality.

### Motivation and Context
I noticed potential for enhancement in the update course form concerning usability and visualization.

### Description
**1) I refactored the form from a template into a reactive form.**
The reason for this was that form validation is more straight forward.
Form validation is currently only showing messages for the short name field, as its requirements are most complicated (I'm not showing messages for required fields, that should be clear from the bootstrap styling)
**2) I refactored the date-picker component into an inline styled one (i saw the todo in the code).**
I had to implement the ControlValueAccess interface so that forms can interact with the component without any tricks (e.g. supplying object and key for value binding).
I went through the code and changed the places where the component was used to make sure everything still works.
**3) I aligned date fields next to each other in a flex container**, because they shouldn't need too much space. I also added a calendar icon for toggling.

### Some further improvements that could be done
- Add cross validation on course start and course end fields in course-update (start needs to be before end, etc)
- Add validation for date in start/end (can not be in past, etc.)

#### Screenshots

![image](https://user-images.githubusercontent.com/28230611/52947156-39c6d400-3376-11e9-8169-86eeec87cd27.png)

![image](https://user-images.githubusercontent.com/28230611/52947176-4814f000-3376-11e9-9035-3664059d94a2.png)

krusche commented:
Another comment: While integrating the latest changes from develop, I noticed that you have deleted the yarn.lock file. Please don't do that as it is recommended to commit this file (see e.g. https://stackoverflow.com/questions/39990017/should-i-commit-the-yarn-lock-file-and-what-is-it-for)

thilo-behnke commented:
Hi Stephan,

Thanks for the review!

About the yarn.lock - my mistake, I only wanted to check it out again from develop because I removed the added dependency for the datepicker again, I must have deleted it instead... sorry!

About the rest:
**1) Positioning of date-picker**
You are right, it was displayed behind the ng bootstrap modal.
The overlay group of the datepicker had a z-index of 1000, while the bootstrap modal has a z-index of 1050.
I added a new rule for the datepicker selector to our global css and commented it appropriately.
Normally it could be a problem to mess around with z-indexes in this way, but because the datepicker always closes when you click another area in the document, it should be fine (the only edge case would be if a modal opens without user input, but this would be bad design anyway).

**2) Validation Regex**
Good find, there was a dollar sign missing at the end, so after 0 or more occurrences of any number/character everything was allowed.

**3) Translations**
I agree, I refactored this part and created a new directive that can translate with dynamic keys.
